### PR TITLE
Added reduce_variance for RaggedTensor with extra optimization

### LINKED ISSUE #37000: Proposal: Modify tf.math.reduce_variance so that it is compatible with ragged tensors
Sorry for filing this under "bug", but the "feature request" submission option didn't show up for me. Please correct this if possible (I copied the template by hand from the markdown file).

**System information**
- TensorFlow version (you are using): 2.1, but applies to master branch as of opening time
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

Currently the variance in the mentioned op is computed as E[(X - E[X])^2]:

https://github.com/tensorflow/tensorflow/blob/c71db95e184447c0008dc007a27d692cd5496d0e/tensorflow/python/ops/math_ops.py#L2112-L2114

Because that requires keepdims == True for the first mean reduction, this is not compatible with ragged tensors. An alternative formulation for the variance could however be E[X**2] - E[X]^2

```
 mean_of_square = reduce_mean(gen_math_ops.square(input_tensor), axis=axis, keepdims=keepdims) 
 square_of_mean = gen_math_ops.square(reduce_mean(input_tensor, axis=axis, keepdims=keepdims)) 
 return mean_of_square - square_of_mean
```

**Will this change the current api? How?**
This might have a reduced numerical stability for Var << Mean, but it could be used at least for ragged tensors in the keepdims == False case. Checking if the tensor is ragged and choosing the implementation based on that would guarantee API stability, but potentially increase maintenance burden later on.

**Who will benefit with this feature?**

Users who want to aggregate statistics over variable-length sequences encoded in ragged tensors.

**Any Other info.**

### PR DESCRIPTION
**Pull request by ashutosh1919**
Fixes #37000. Please take a look at [this proof](https://math.stackexchange.com/a/755344) for the formula used. So, it is good to modify this if  it's optimization as suggested in the issue.

### CODE REVIEW

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]
This would result in numerical stability changes.

The linked issue mentions doing this computation only for ragged tensors, not everywhere.

**Comment by alexrobomind** [comment]
Just to clarify, the purpose of this suggestion was not to improve performance, but for compatibility with ragged tensors. In fact, the current method is 2x mean, 1x square, 1x diff, while the proposed one is 2x mean, 2x square, 2x diff, making it one op more. So it would probably be best to only apply this method to ragged tensors.

**Comment by gbaned** [comment]
@ashutosh1919 Any update on this PR, please. Thanks!

**Comment by ashutosh1919** [comment]
@mihaimaruseac , Added condition to check instance of `RaggedTensor`.

File: tensorflow/python/ops/math_ops.py [hunk 1]
```diff
@@ -2109,6 +2110,14 @@ def reduce_variance(input_tensor, axis=None, keepdims=False, name=None):
   """
   name = name if name else "reduce_variance"
   with ops.name_scope(name):
+    if isinstance(input_tensor, ragged_tensor.RaggedTensor):
+      square_of_input = gen_math_ops.square(input_tensor)
+      mean_of_square = reduce_mean(square_of_input,
+      	                           axis=axis,
+      	                           keepdims=keepdims)
```
Comment (mihaimaruseac) [inline tensorflow/python/ops/math_ops.py]: Make sure you run pylint, the indentation on these lines looks strange

**Review by mihaimaruseac — DISMISSED** [review]

File: tensorflow/python/ops/math_ops.py (hunk 1 above)
Comment (ashutosh1919) [reply tensorflow/python/ops/math_ops.py]: done

**Review by mihaimaruseac — APPROVED** [review]

**Comment by gbaned** [comment]
@ashutosh1919 Can you please address Ubuntu Sanity errors? Thanks!

**Comment by ashutosh1919** [comment]
@gbaned , I don't think this error is because of the code that I changed in this PR. But I will rebase master so that If it has to do something with that then it will be resolved.

**Review by mihaimaruseac — DISMISSED** [review]

**Comment by gbaned** [comment]
@ashutosh1919 Still, Ubuntu Sanity errors appearing, Can you fix those ? Thanks!

**Comment by ashutosh1919** [comment]
@gbaned, as I mentioned in the previous comment that this error is not because of the code which I added. And if it is because of that then I am not able to see the error messages corresponding to that.

**Comment by mihaimaruseac** [comment]
Actually, the error is from this PR, it is a pylint error.

```
FAIL: Found 2 non-whitelisted pylint errors:
tensorflow/python/ops/math_ops.py:2116: [C0330(bad-continuation), ] Wrong continued indentation (add 7 spaces).

tensorflow/python/ops/math_ops.py:2117: [C0330(bad-continuation), ] Wrong continued indentation (add 7 spaces).
```

File: tensorflow/python/ops/math_ops.py [hunk 2]
```diff
@@ -2109,6 +2110,14 @@ def reduce_variance(input_tensor, axis=None, keepdims=False, name=None):
   """
   name = name if name else "reduce_variance"
   with ops.name_scope(name):
+    if isinstance(input_tensor, ragged_tensor.RaggedTensor):
+      square_of_input = gen_math_ops.square(input_tensor)
+      mean_of_square = reduce_mean(square_of_input,
+      	                     axis=axis,
+      	                     keepdims=keepdims)
```
Comment (mihaimaruseac) [inline tensorflow/python/ops/math_ops.py]: Make sure you don't mix tabs and spaces here. Visually it looks like it has the proper indentation, but pylint complains asking to insert a few more spaces.

**Comment by ashutosh1919** [comment]
Apologies @mihaimaruseac and @gbaned , I didn't notice that silly mistake. I have resolved that. Please run ci. Let's check if it works now.

**Review by mihaimaruseac — APPROVED** [review]

**Comment by mihaimaruseac** [comment]
There are slightly under 4000 tests failing internally with this PR. We will have to import it manually

**Comment by ashutosh1919** [comment]
@mihaimaruseac , Please suggest the chnages. I am facing this issue for the first time and that's why don't know much about it.

**Comment by mihaimaruseac** [comment]
There are internal tests which depend on the numerical precision of the old code. So far I don't think you have to do anything

**Comment by gbaned** [comment]
@mihaimaruseac Any update on this PR, please. Thanks!

**Comment by gbaned** [comment]
@mihaimaruseac Any update on this PR, please. Thanks!

**Comment by gbaned** [comment]
@mihaimaruseac Any update on this PR, please. Thanks!

**Comment by gbaned** [comment]
@mihaimaruseac Any update on this PR, please. Thanks!

**Comment by gbaned** [comment]
@mihaimaruseac Any update on this PR, please. Thanks!

**Comment by gbaned** [comment]
@mihaimaruseac Any update on this PR, please. Thanks!

**Comment by gbaned** [comment]
@mihaimaruseac Any update on this PR, please. Thanks!

**Comment by edloper** [comment]
Instead of adding a special case in reduce_variance in math_ops.py, this should be implemented by defining a `reduce_variance` method in `python/ops/ragged/ragged_math_ops.py`, and registering it for type-based dispatch in `python/ops/ragged/ragged_dispatch.py`.  (You can search for `reduce_mean` in those two files, and it should be fairly clear what needs to be done.)  Ideally, you should also add tests in `ragged_reduce_op_test.py` and `ragged_dispatch_test.py` as well.

**Comment by gbaned** [comment]
@ashutosh1919 Can you please check @edloper's comments and keep us posted ? Thanks!

**Comment by gbaned** [comment]
@ashutosh1919 Any update on this PR? Please. Thanks!

**Comment by ashutosh1919** [comment]
@edloper , @mihaimaruseac - I have added reduce_variance and its tests as per previous comments by @edloper . Please review.

File: tensorflow/python/ops/ragged/ragged_math_ops.py [hunk 3]
```diff
@@ -630,6 +637,18 @@ def reduce_mean(input_tensor, axis=None, keepdims=None, name=None):
       return total / count
 
 
+def reduce_variance(input_tensor, axis=None, keepdims=False, name=None):
+  """For docs, see: _RAGGED_REDUCE_DOCSTRING."""
+  with ops.name_scope(name, 'RaggedReduceVariance', [input_tensor, axis]):
+    square_of_input = gen_ragged_math_ops.square(input_tensor)
```
Comment (edloper) [inline tensorflow/python/ops/ragged/ragged_math_ops.py:643]: This should be `math_ops.square`, not `gen_ragged_math_ops.square`.  (Are the tests passing?  I would have expected this to be caught by the tests.)  (Similarly, line 648 below should be changed to `math_ops.square`.)

File: tensorflow/python/ops/ragged/ragged_reduce_op_test.py [hunk 4]
```diff
@@ -39,6 +39,16 @@ def mean(*values):
   return 1.0 * sum(values) / len(values)
 
 
+def variance(*values):
```
Comment (edloper) [inline tensorflow/python/ops/ragged/ragged_reduce_op_test.py:42]: Suggestion: have this just return `np.var(values)`.  (Your implementation looks right to me, but if we use a standard/reference implementation then it's one less thing to think about.)

File: tensorflow/python/ops/ragged/ragged_math_ops.py (hunk 3 above)
Comment (ashutosh1919) [reply tensorflow/python/ops/ragged/ragged_math_ops.py:643]: done

File: tensorflow/python/ops/ragged/ragged_reduce_op_test.py (hunk 4 above)
Comment (ashutosh1919) [reply tensorflow/python/ops/ragged/ragged_reduce_op_test.py:42]: done

**Comment by ashutosh1919** [comment]
@edloper , apologies. Actually i was running the bazel tests in my local in parallel. But it failed at the end and I noticed that today morning. 
I have done changes that you mentioned and now bazel tests are passing in my local.
Please review.

**Review by bhack — APPROVED** [review]
Just to run Kokoro

**Comment by ashutosh1919** [comment]
Resolved doctest failiure. But there was also error shown as below.
```
AttributeError: module 'tensorflow.compat.v2' has no attribute 'reduce_variance'
```

@bhack , can you please run the kokoro?

**Comment by bhack** [comment]
isn't `tf.math.reduce_variance`?

**Comment by ashutosh1919** [comment]
@bhack , Thanks for noticing the error. Done changes. Hope it'll work now.

**Review by bhack — APPROVED** [review]
Just for kokoro

**Review by mihaimaruseac — APPROVED** [review]

**Comment by ashutosh1919** [comment]
@mihaimaruseac , can you please approve again? Fixed doctest

**Review by bhack — APPROVED** [review]
For kokoro

**Comment by ashutosh1919** [comment]
@mihaimaruseac , can you please look at 3 failed builds? Looks like it is not because of this PR.

File: tensorflow/python/ops/ragged/ragged_math_ops.py [hunk 5]
```diff
@@ -630,6 +637,18 @@ def reduce_mean(input_tensor, axis=None, keepdims=None, name=None):
       return total / count
 
 
+def reduce_variance(input_tensor, axis=None, keepdims=False, name=None):
+  """For docs, see: _RAGGED_REDUCE_DOCSTRING."""
+  with ops.name_scope(name, 'RaggedReduceVariance', [input_tensor, axis]):
+    square_of_input = gen_math_ops.square(input_tensor)
```
Comment (edloper) [inline tensorflow/python/ops/ragged/ragged_math_ops.py:643]: Using `gen_math_ops.square` works here, but it's preferable to just use `math_ops.square`.  (`gen_math_ops.py` is the generated python stub file, and all of its symbols are imported into `math_ops.py`.)

**Review by edloper — APPROVED** [review]
Thanks for adding this!

**Comment by ashutosh1919** [comment]
@edloper , can you please approve again? I changed `gen_math_ops` to `math_ops`

File: tensorflow/python/ops/ragged/ragged_math_ops.py (hunk 5 above)
Comment (ashutosh1919) [reply tensorflow/python/ops/ragged/ragged_math_ops.py]: done the change

File: tensorflow/python/ops/ragged/ragged_math_ops.py [hunk 6]
```diff
@@ -25,6 +25,7 @@
 from tensorflow.python.framework import tensor_util
 from tensorflow.python.ops import array_ops
 from tensorflow.python.ops import check_ops
+from tensorflow.python.ops import gen_math_ops
```
Comment (edloper) [inline tensorflow/python/ops/ragged/ragged_math_ops.py:28]: The lint checker may complain that this import is now unused.

**Review by edloper — APPROVED** [review]

File: tensorflow/python/ops/ragged/ragged_math_ops.py (hunk 6 above)
Comment (ashutosh1919) [reply tensorflow/python/ops/ragged/ragged_math_ops.py]: Removed. Can you please approve again @edloper ?

**Review by edloper — APPROVED** [review]

**Pull request merged**